Mathlib/Computability/PrimrecFilter.lean

```diff
@@ -0,0 +1,209 @@
+    [DecidableRel f] (hf : PrimrecRel f) :
+    PrimrecRel fun (L : List α) => fun b => (∀ a ∈ L, f a b) := by
+  let g (b : β) := fun L => List.filter (fun a => f a b) L
+  have h (L : List α) (b : β) : (g b L).length = L.length ↔ (∀ a ∈ L, f a b) := by simp [g]
+  apply PrimrecRel.of_eq ?_ h
+  unfold PrimrecRel Primrec₂
+  simp only
+  rw [← PrimrecPred]
+  refine PrimrecRel.comp Primrec.eq ?_ (Primrec.comp list_length fst)
+  apply Primrec.comp list_length
+  refine Primrec₂.comp ?_ snd fst
+  apply Primrec₂.swap
+  exact primrec₂.list_filter f hf
+
+/-- If f a b is decidable, then for any fixed n and y,  "g n y ↔ ∃ x < n, f x y" is a
+primitive recursive relation. -/
+lemma bounded_exists (f : ℕ → ℕ → Prop) [DecidableRel f]
+    (hf : PrimrecRel f) : PrimrecRel (fun n => (fun y => (∃ x < n, f x y))) := by
+  have h : PrimrecRel (fun n => (fun y => (∃ x ∈ range n, f x y))) := by
+    apply PrimrecRel.comp (filter_exists f hf) (Primrec.comp list_range fst) snd
+  apply PrimrecRel.of_eq h
+  simp
+
+/-- If f a b is decidable, then for any fixed n and y,  "g n y ↔ ∀ x < n, f x y" is a
+primitive recursive relation. -/
+lemma bounded_forall (f : ℕ → ℕ → Prop) [DecidableRel f]
+    (hf : PrimrecRel f) : PrimrecRel (fun n => (fun y => (∀ x < n, f x y))) := by
+  have h : PrimrecRel (fun n => (fun y => (∀ x ∈ range n, f x y))) := by
+    apply PrimrecRel.comp (filter_forall f hf) (Primrec.comp list_range fst) snd
+  apply PrimrecRel.of_eq h
+  simp
+
+end PrimrecRel
+
+namespace Primrec
+
+/-- A helper lemma for proofs about bounded quantifiers on decidable relations. -/
+lemma nat_rel_list_filter (f : ℕ → ℕ → Prop) (s : ℕ) [DecidableRel f] (hf : PrimrecRel f) :
+    Primrec fun n => ((List.range (s)).filter (fun y => f y n)) := by
+  let g (n : ℕ): ℕ → Option Nat := (fun y => (if f y n = True then y else Option.none))
+  have h (n : ℕ): (range (s)).filter (fun y => f y n) = filterMap (g n) (List.range s) := by
+    simp [g]
+    rw [← List.filterMap_eq_filter]
+    refine filterMap_congr ?_
+    simp [Option.guard]
+  simp [h]
+  apply listFilterMap
+  · exact Primrec.const (range s)
+  · apply Primrec.ite
+    · simp
+      exact PrimrecRel.comp hf snd fst
+    · exact option_some_iff.mpr snd
+    · exact Primrec.const Option.none
+
+end Primrec
+
+namespace PrimrecPred
+
+/-- If f a b is decidable, then for any fixed n and y, "∃ x < n, f x y" is a
+primitive recursive predicate in n. This is sometimes easier to work with than the fully
+general case involving a primitive recursive relation. -/
+lemma bounded_exists (f : ℕ → ℕ → Prop) (s : ℕ) [DecidableRel f]
+    (hf : PrimrecRel f) :
+    PrimrecPred (fun n => ∃ y < s, (f y n)) := by
+  have h1 : (fun n => decide (∃ y < s, f y n)) =
+            (fun n => decide ((List.range s).filter (fun y => f y n) ≠ [])) := by simp
+  simp [PrimrecPred, h1]
+  apply PrimrecPred.not
+  apply PrimrecRel.comp Primrec.eq ?_ (const [])
+  apply nat_rel_list_filter
```

Comment:
Can this become e.g. `nat_rel_list_filter hf`? If so, please do so.

Reply:
Does `nat_rel_list_filter _ _ hf` work? (Perhaps with some more or fewer underscores.)